lib/ex_doc/formatter/html/autolink.ex

```diff
@@ -142,32 +154,86 @@ defmodule ExDoc.Formatter.HTML.Autolink do
-        if source = get_source(alias, aliases, lib_dirs) do
-          n = enc_h("#{name}")
-          {string_to_link, string_with_parens} = split_string_to_link(string)
-          ~s[<a href="#{source}#{enc_h(inspect alias)}.html#t:#{n}/#{length(args)}">#{h(string_to_link)}</a>#{string_with_parens}]
-        else
-          string
-        end
-      _, string ->
-        string
-    end)
+    ref = make_ref()
+
+    {ast, placeholders} =
+      Macro.prewalk(ast, %{}, fn
+        {:::, _, [{name, meta, args}, right]}, placeholders when is_atom(name) and is_list(args) ->
+          {{:::, [], [{{ref, name}, meta, args}, right]}, placeholders}
+
+        # Consume this form so that we don't autolink `foo` in `foo :: bar`
+        {{^ref, name}, _, args}, placeholders when is_atom(name) and is_list(args) ->
+          {{name, [], args}, placeholders}
+
+        {name, _, args} = form, placeholders when is_atom(name) and is_list(args) ->
+          arity = length(args)
+
+          if {name, arity} in typespecs do
+            string = Macro.to_string(form)
+            n = enc_h("#{name}")
+            {string_to_link, _string_with_parens} = split_string_to_link(string)
+            string = ~s[<a href="#t:#{n}/#{arity}">#{h(string_to_link)}</a>]
+
+            put_placeholder(form, string, placeholders)
+          else
+            {form, placeholders}
+          end
+
+        {{:., _, [alias, name]}, _, args} = form, placeholders when is_atom(name) and is_list(args) ->
+          alias = expand_alias(alias)
+
+          if source = get_source(alias, aliases, lib_dirs) do
+            string = Macro.to_string(form)
+            n = enc_h("#{name}")
+            {string_to_link, _string_with_parens} = split_string_to_link(string)
+            string = ~s[<a href="#{source}#{enc_h(inspect alias)}.html#t:#{n}/#{length(args)}">#{h(string_to_link)}</a>]
+
+            put_placeholder(form, string, placeholders)
+          else
+            {form, placeholders}
+          end
+
+        form, placeholders ->
+          {form, placeholders}
+      end)
+
+    ast
+    |> format_ast()
+    |> replace_placeholders(placeholders)
   end
 
-  defp short_typespec?(ast) do
-    byte_size(Macro.to_string(ast)) <= 70
+  defp put_placeholder(form, string, placeholders) do
+    id = zero_pad(map_size(placeholders), 3)
+    placeholder = :"_p#{id}_"
+    form = put_elem(form, 0, placeholder)
+    {form, Map.put(placeholders, Atom.to_string(placeholder), string)}
+  end
+
+  defp zero_pad(val, count) do
+    num = Integer.to_string(val)
+    pad_length = max(count - byte_size(num), 0)
+    :binary.copy("0", pad_length) <> num
+  end
+
+  defp replace_placeholders(string, placeholders) do
+    Regex.replace(~r"_p\d{3}_", string, &Map.fetch!(placeholders, &1))
+  end
+
+  defp format_ast(ast) do
+    string = Macro.to_string(ast)
+
+    if formatter_available?() do
+      string
+      |> Code.format_string!(line_length: @line_length)
```

Comment:
for line_length: 70 I noticed it caused churn in a few places, e.g.: https://hexdocs.pm/elixir/master/Access.html#elem/1 vs:

<img width="854" alt="zrzut ekranu 2017-11-30 o 01 24 31" src="https://user-images.githubusercontent.com/76071/33406774-c7dbf82c-d56e-11e7-8471-0dde89739a26.png">

It might be better to stick to default line length on formatter (which then breaks other tests and I didn't want to necessarily add more `if formatter_available?` clauses or adjust tests 😅 )

Reply:
Let's stick with 80, yeah?